src/metabase/util.clj

```diff
@@ -959,3 +959,24 @@
   non-unique at times, numeric ids are not guaranteed."
   [field-ref]
   (into [] (take 2) field-ref))
+
+(def preserved-keys
+  "Keys that can survive merging metadata from the database onto metadata computed from the query. When merging
+  metadata, the types returned should be authoritative. But things like semantic_type, display_name, and description
+  can be merged on top."
+  ;; TODO: ideally we don't preserve :id but some notion of :user-entered-id or :identified-id
+  [:id :description :display_name :semantic_type :fk_target_field_id :settings])
+
+(defn combine-metadata
+  "Blend saved metadata from previous runs into fresh metadata from an actual run of the query.
+
+  Ensure that saved metadata from datasets or source queries can remain in the results metadata. We always recompute
+  metadata in general, so need to blend the saved metadata on top of the computed metadata. First argument should be
+  the metadata from a run from the query, and `pre-existing` should be the metadata from the database we wish to
+  ensure survives."
+  [fresh pre-existing]
+  (let [by-key (key-by (comp field-ref->key :field_ref) pre-existing)]
+    (for [{:keys [field_ref] :as col} fresh]
+      (if-let [existing (get by-key (field-ref->key field_ref))]
+        (merge col (select-keys existing preserved-keys))
+        col))))
```

Comment:
I think this stuff is too niche to go in `metabase.util`. 

The way I look at is is that if you can kind of divide Metabase up into a few separate conceptual modules like the Query Processor/Driver Interface, models layer, web server/REST API handlers/etc. Things in `metabase.util` are basically utility functions that are shared by all the different modules; it's basically like our own personal [Medley](https://github.com/weavejester/medley) library. 

These functions are not really conceptually part of anything except the QP/driver stuff -- I actually think maybe you should just stick everything related to calculating query metadata into its own QP namespace including these functions and `result-metadata-async` as well. Figuring out how to calculate results metadata shouldn't be a concern for anyone outside of the QP module

Reply:
you are right. Good catch